File: python/Q139_word_break.py
#
# @lc app=leetcode id=139 lang=python3
#
# [139] Word Break
#


# @lc code=start
class Solution:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        dp: list[bool] = [False] * len(s)
        sorted_word_dict: list[str] = sorted(wordDict, key=lambda w: len(w))

        for i in range(len(s)):
            for word in sorted_word_dict:
                if (
                    i + 1 - len(word) == 0 or (i >= len(word) and dp[i - len(word)])
                ) and s[i - len(word) + 1 : i + 1] == word:
                    dp[i] = True
                    break

        return dp[-1]


# @lc code=end

File: rust/src/leetcode/Q139_word_break.rs
/*
 * @lc app=leetcode id=139 lang=rust
 *
 * [139] Word Break
 */

// @lc code=start
impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut dp: Vec<bool> = vec![false; s.len() + 1];
        dp[0] = true;

        for i in 1..=s.len() {
            for word in &word_dict {
                if word.len() <= i && s[i - word.len()..i] == word[..] && dp[i - word.len()] {
                    dp[i] = true;
                    break;
                }
            }
        }

        dp[dp.len() - 1]
    }
}
// @lc code=end
